### DISCUSSION TITLE
About '-fwasm-exceptions' to improve performance

### DISCUSSION THREAD

**Discussion by frank-pian** [donalffons/opencascade.js#122] [General]
"-fexceptions" causes significant performance degradation. When I disabled exceptions, the efficiency of testing 'ReadStep' improved significantly and was close to that of native C++.
However, the following discussion disabling exceptions is not recommended and can lead to logic errors.
[Exception "Geom_BSplineSurface: Weights values too small" when reading STEP file](https://dev.opencascade.org/content/exception-geombsplinesurface-weights-values-too-small-when-reading-step-file)
So I tried the '-fwasm-exceptions' compilation parameter as recommended. This resulted in a large number of errors leading to compilation crashes.
```
fatal error: error in backend: In function _ZNK38GeomLib_CheckCurveOnSurface_TargetFunc5ValueEdRd: When using Wasm EH with Emscripten SjLj, there is a restriction that `setjmp` function call and exception cannot be used within the same function:
  %11 = invoke noundef ptr @_ZN21Standard_ErrorHandlerC1Ev(ptr noundef nonnull align 4 dereferenceable(176) %7)
          to label %12 unwind label %109
```
There are cases in OCCT code where setjmp and exception are used at the same time, so should we try to remove setjmp function?

**Comment by donalffons** [comment]
Yes, the performance degradation and file size impact is currently massive (as I mentioned in the thread over at the OpenCascade forums) when using non-native exceptions. Using native wasm exceptions, the performance & size impact of enabling exceptions is significantly lower. My opinion is to not enable native exception support in OpenCascade.js until there is stable support in current versions of Chromium, Firefox and Safari. Firefox was lacking support until very recently.

If you want to enable native wasm support using `-fwasm-exceptions`, you would have to re-build the entire Docker image using this flag, since native wasm support has to be enabled at compile time (and then use that docker image for linking in the docker - run phase). The docker run - phase performs mainly linking using the pre-compiled OpenCascade assets (see [this diagram](https://ocjs.org/docs/developer-docs/overview)). You would probably have to search-and-replace the code for occurences of [`-fexceptions`](https://github.com/donalffons/opencascade.js/blob/master/src/customBuildSchema.py#L28) and replace those with `-fwasm-exceptions`. Be prepared for extremely long build times when building the Docker image, though.

However, it seems like the situation has changed. [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WebAssembly/Exception) says that Firefox now has "Preview browser support" for that feature, so I think it shouldn't be long until we can make the switch in OpenCascade.js.

Depending on the application you're building, you might not need exception support. My opinion is that if you don't need exception support, you should (at least try to) disable it in your production application as this greatly improves size & performance, regardless of whether native or non-native exceptions are used. Exception support is only needed (I believe) if
- you're using OpenCascade code that both throws and catches exceptions, or
- if you need to inspect thrown exceptions in JS

**Reply by donalffons** [reply]
I made some performance benchmarks a while ago with native exceptions using Chrome [in this thread](https://github.com/WebAssembly/exception-handling/issues/20). This might give you an idea about the file size difference between native and non-native exception support. The "Aw, Snap!" errors I had seem to be resolved now (just tested it in Chrome 101.0.4951.54). So that's pretty great news.

**Comment by frank-pian** [comment]
I tried to replace all the '-fexceptions'. The following error occurred while building the Docker Image.
```
DISABLE_EXCEPTION_THROWING was set (probably from -fno-exceptions) but is not compatible with enabling exception catching (DISABLE_EXCEPTION_CATCHING=0). If you don't want exceptions, set DISABLE_EXCEPTION_CATCHING to 1; if you do want exceptions, don't link with -fno-exceptions
```
But I didn't find DISABLE_EXCEPTION_THROWING.

**Reply by donalffons** [reply]
That error message (from Emscripten) is weird and seems misleading. Removing `-sDISABLE_EXCEPTION_CATCHING=0` from the code seems to fix this. I made a [branch](https://github.com/donalffons/opencascade.js/tree/native-wasm-exceptions) with native exception support and am running a build right now, which will publish an image to Dockerhub. Will post an update here, once it has gone through.

**Reply by donalffons** [reply]
There is now a single-threaded Docker image with support for native wasm exceptions under the tag [donalffons/opencascade.js:staging-native-wasm-exceptions](https://hub.docker.com/layers/212806812/donalffons/opencascade.js/staging-native-wasm-exceptions/images/sha256-9aca1f5439d402f4bb3ad4d03eee2365033a5b12735fc4bbb19b5a9ace568dc9?context=repo).

But I haven't yet managed to actually catch and extract any information from an exception. In [this test](https://github.com/donalffons/opencascade.js/blob/master/test/index.test.ts#L276), the exception is of type [`WebAssembly.Exception`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WebAssembly/Exception) - as opposed to a `number` in the non-native case... So, currently that test fails...

**Reply by donalffons** [reply]
I started [this thread](https://github.com/emscripten-core/emscripten/discussions/16946) over at Emscripten, asking on to extract information from a native wasm-exception.

**Comment by frank-pian** [comment]
I want to start with the official Opencascade cmake build. Do you have any experience in this field ?
I built the method according to the official cmake and replaced '-fexceptions', but something went wrong.
https://github.com/frank-pian/occt_wasm_exception_test/blob/main/Dockerfile
Thanks a million !

**Reply by donalffons** [reply]
I haven't looked at your Dockerfile in detail...
Version 1.1.1 of OpenCascade.js was using OCCT's cmake build system also. Maybe it helps to take a look at [this cmake command](https://github.com/donalffons/opencascade.js/blob/v1.1.1/make.py#L228) or [this emcc command](https://github.com/donalffons/opencascade.js/blob/v1.1.1/make.py#L319)?!

**Reply by frank-pian** [reply]
That seems to help a lot. I didn't find 'OCC_CONVERT_SIGNALS' in your project build parameters. So it is possible not to use setjump function.
Thank you again for.
